geoutils/filters.py

```diff
@@ -0,0 +1,296 @@
+
+    :param array: The input array to be filtered.
+    :param size:  the shape that is taken from the input array, at every element position,
+    to define the input to the filter function
+
+    :returns: The filtered array (same shape as input).
+    """
+    # Check that array dimension is 2 or 3
+    if np.ndim(array) not in [2, 3]:
+        raise ValueError(f"Invalid array shape given: {array.shape}. Expected 2D or 3D array.")
+
+    nans = np.isnan(array)
+    # We replace temporarily NaNs by infinite values during filtering to avoid spreading NaNs
+    array_nans_replaced = np.where(nans, np.inf, array)
+    array_nans_replaced_f = scipy.ndimage.minimum_filter(array_nans_replaced, size=size, **kwargs)
+    # In the end, we want the filtered array without infinite values, so we put back NaNs
+    return np.where(nans, array, array_nans_replaced_f)
+
+
+def max_filter(array: NDArrayNum, size: int = 5, **kwargs: Any) -> NDArrayNum:
+    """
+    Apply a maximum filter to a raster that may contain NaNs, using scipy's implementation.
+
+    :param array: the input array to be filtered.
+    :param size:  the shape that is taken from the input array, at every element position,
+    to define the input to the filter function
+
+    :returns: the filtered array (same shape as input).
+    """
+    # Check that array dimension is 2 or 3
+    if np.ndim(array) not in [2, 3]:
+        raise ValueError(f"Invalid array shape given: {array.shape}. Expected 2D or 3D array.")
+
+    nans = np.isnan(array)
+    # We replace temporarily NaNs by negative infinite values during filtering to avoid spreading NaNs
+    array_nans_replaced = np.where(nans, -np.inf, array)
+    array_nans_replaced_f = scipy.ndimage.maximum_filter(array_nans_replaced, size=size, **kwargs)
+    # In the end we want the filtered array without negative infinite values, so we put back NaNs
+    return np.where(nans, array, array_nans_replaced_f)
+
+
+def distance_filter(array: NDArrayNum, radius: float, outlier_threshold: float) -> NDArrayNum:
+    """
+    Filter out pixels whose value is distantly more than a set threshold from the average value of all neighbor \
+pixels within a given radius.
+    Filtered pixels are set to NaN.
+
+    TO DO: Add an option on how the "average" value should be calculated, i.e. using a Gaussian, median etc filter.
+
+    :param array: the input array to be filtered.
+    :param radius: the radius in which the average value is calculated (for Gaussian filter, this is sigma).
+    :param outlier_threshold: the minimum difference abs(array - mean) for a pixel to be considered an outlier.
+
+    :returns: the filtered array (same shape as input)
+    """
+    # Create mask of valid (finite) values
+    valid_mask = np.isfinite(array)
+
+    # Smooth both the data and the valid mask
+    smoothed = gaussian_filter(np.nan_to_num(array, nan=0.0), sigma=radius)
+    normalization = gaussian_filter(valid_mask.astype(float), sigma=radius)
```

Comment:
Ok, so the fix for taking into account nans is used here. Why not directly in the main `gaussian_filter`?